src/terminal/adapter/FontBuffer.cpp

```diff
@@ -0,0 +1,603 @@
+            return { 6, 10, 0 };
+        }
+        else if (inRange(9, 12))
+        {
+            // VT320 - 9x12 cell with default text width.
+            return { 9, 12, textWidth };
+        }
+        else if (inRange(6, 16))
+        {
+            // VT420 & VT5xx - 6x16 cell with default text width.
+            return { 6, 16, textWidth };
+        }
+        else if (inRange(6, 20))
+        {
+            // VT340 - 6x20 cell with default text width.
+            return { 6, 20, textWidth };
+        }
+        else if (inRange(7, 30))
+        {
+            // VT382 - 7x30 cell with default text width.
+            return { 7, 30, textWidth };
+        }
+        else
+        {
+            // If all else fails, assume the maximum size.
+            return { MAX_WIDTH, MAX_HEIGHT, textWidth };
+        }
+    }
+}
+
+void FontBuffer::_packAndCenterBitPatterns()
+{
+    // If this is a text font, we'll clip the bits up to the text width and
+    // center them within the full cell width. For a full cell font we'll just
+    // use all of the bits, and no offset will be required.
+    _textWidth = _textWidth ? _textWidth : _fullWidth;
+    _textWidth = std::min(_textWidth, _fullWidth);
+    _textOffset = (_fullWidth - _textWidth) / 2;
+    const auto textClippingMask = ~(0xFFFF >> _textWidth);
+
+    // If the text is given an explicit width, we check to what extent the
+    // content is offset from center. Knowing that information will enable the
+    // renderer to scale the font more symmetrically.
+    _textCenteringHint = _declaredWidth ? _fullWidth - (_declaredWidth + _textOffset * 2) : 0;
+
+    // Initially the characters are written to the buffer assuming the maximum
+    // cell height, but now that we know the true height, we need to pack the
+    // buffer data so that each character occupies the exact number of scanlines
+    // that are required.
+    for (auto srcLine = 0u, dstLine = 0u; srcLine < _buffer.size(); srcLine++)
+    {
+        if ((srcLine % MAX_HEIGHT) < _fullHeight)
+        {
+            auto characterScanline = til::at(_buffer, srcLine);
+            characterScanline &= textClippingMask;
+            characterScanline >>= _textOffset;
+            til::at(_buffer, dstLine++) = characterScanline;
+        }
+    }
+}
+
+void FontBuffer::_fillUnusedCharacters()
+{
+    // Every character in the buffer that hasn't been uploaded will be replaced
+    // with an error glyph (a reverse question mark). This includes every
+    // character prior to the start char, or after the last char.
+    const auto errorPattern = _generateErrorGlyph();
+    for (auto ch = 0u; ch < MAX_CHARS; ch++)
+    {
+        if (ch < _startChar || ch > _lastChar)
+        {
+            auto charBuffer = std::next(_buffer.begin(), ch * _fullHeight);
+            std::copy_n(errorPattern.begin(), _fullHeight, charBuffer);
+        }
+    }
+}
+
+auto FontBuffer::_generateErrorGlyph() -> std::array<uint16_t, MAX_HEIGHT>

```

Comment:
Yeah. I suspect there might have been a reason for it in an earlier draft of the code that became unnecessary as things evolved. I'll see if I can remote it.

Reply:
FYI, I've updated this now. I think the reason I was using it was to avoid the class scope prefix on the `MAX_HEIGHT` constant. But considering that saved all of 4 characters, it really doesn't seem like it was worth the confusion.